src/net/socket/udp_sock.c

```diff
@@ -46,43 +49,65 @@ static int udp_sendmsg(struct sock *sk, struct msghdr *msg, int flags) {
 	assert(msg->msg_iov);
 	assert(msg->msg_iov->iov_base);
 
-	data_len = msg->msg_iov->iov_len;
-	total_len = actual_len = UDP_HEADER_SIZE + data_len;
+	data_len = 0;
+	for (i = 0; i < msg->msg_iovlen; i++) {
+		data_len += msg->msg_iov[i].iov_len;
+	}
+	data_len_ = data_len;
+
+	actual_len = UDP_HEADER_SIZE + data_len;
+
 	skb = NULL;
 	sockaddr = (const struct sockaddr *)msg->msg_name;
 
-	ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
-	if (ret != 0) {
-		return ret;
-	}
+	while (data_len_ > 0) {
+		ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
 
-#if 0
-	if (actual_len < total_len) {
-		skb_free(skb);
-		return -EMSGSIZE;
-	}
-#endif
+		if (ret != 0) {
+			return ret;
+		}
 
-	if (msg->msg_name != NULL) {
-		to = (const struct sockaddr_in *)msg->msg_name;
-	} else {
-		to = (const struct sockaddr_in *)&to_inet_sock(sk)->dst_in;
-	}
+		if (msg->msg_name != NULL) {
+			to = (const struct sockaddr_in *)msg->msg_name;
+		} else {
+			to = (const struct sockaddr_in *)&to_inet_sock(sk)->dst_in;
+		}
+
+		assert(skb);
+		assert(skb->h.uh);
+
+		udp_header = (unsigned char *) (skb->h.uh + 1);
+		size = sizeof(msg->msg_iov[i].iov_base[0]);
 
-	assert(skb);
-	assert(skb->h.uh);
+		iov_len = i = skb_len = 0;
 
-	udp_build(skb->h.uh, sock_inet_get_src_port(sk), to->sin_port, total_len);
+		for (i = 0; i < msg->msg_iovlen; i++) {
+			iov_len = msg->msg_iov[i].iov_len * size;
```

Comment:
read/write (and readv/wirtev) operates on bytes level, in contrast with stdlib fread/fwrite. This line may confuse the reader, so I'd suggest just delete the size and don't scale here